tests/lsp/test_progress.py

```diff
@@ -15,53 +15,95 @@
     def __init__(self):
         super().__init__()
         self.client.notifications: List[ProgressParams] = []
+        self.client.method_calls: List[WorkDoneProgressCreateParams] = []
 
         @self.server.feature(
             CODE_LENS,
-            CodeLensOptions(resolve_provider=False,
-                            work_done_progress=PROGRESS_TOKEN),
+            CodeLensOptions(resolve_provider=False, work_done_progress=True),
         )
-        def f1(params: CodeLensParams) -> Optional[List[CodeLens]]:
+        async def f1(params: CodeLensParams) -> Optional[List[CodeLens]]:
+            if "client_initiated_token" in params.text_document.uri:
+                token = params.work_done_token
+            else:
+                assert "server_initiated_token" in params.text_document.uri
+                token = params.text_document.uri[len("file://") :]
+                if "async" in params.text_document.uri:
+                    await self.server.progress.create_async(token)
+                else:
+                    f = self.server.progress.create(token)
+                    await asyncio.sleep(0.1)
+                    f.result()
+
+            assert token
             self.server.lsp.progress.begin(
-                PROGRESS_TOKEN, WorkDoneProgressBegin(
-                    title="starting", percentage=0)
-            )
-            self.server.lsp.progress.report(
-                PROGRESS_TOKEN,
-                WorkDoneProgressReport(message="doing", percentage=50),
-            )
-            self.server.lsp.progress.end(
-                PROGRESS_TOKEN, WorkDoneProgressEnd(message="done")
+                token,
+                WorkDoneProgressBegin(title="starting", percentage=0),
             )
+            await asyncio.sleep(0.1)
+            if self.server.lsp.progress.tokens[token].cancelled():
+                self.server.lsp.progress.end(
+                    token, WorkDoneProgressEnd(message="cancelled")
+                )
+            else:
+                self.server.lsp.progress.report(
+                    token,
+                    WorkDoneProgressReport(message="doing", percentage=50),
+                )
+                self.server.lsp.progress.end(token, WorkDoneProgressEnd(message="done"))
             return None
 
         @self.client.feature(PROGRESS_NOTIFICATION)
         def f2(params):
             self.client.notifications.append(params)
+            if params.value["kind"] == "begin" and "cancel" in params.token:
+                # client cancels the progress token
+                self.client.lsp.notify(
+                    WINDOW_WORK_DONE_PROGRESS_CANCEL,
+                    WorkDoneProgressCancelParams(token=params.token),
+                )
+
+        @self.client.feature(WINDOW_WORK_DONE_PROGRESS_CREATE)
+        def f3(params: WorkDoneProgressCreateParams):
+            self.client.method_calls.append(params)
+            # XXX return a WorkDoneProgressCreateParams for now,
```

Comment:
What does XXX mean here? Like "TODO"?

Reply:
Yes, something like that. I wanted to leave a note explaining that this is a workaround for that problem. For now it's not really actionnable until that problem with types of LSP_METHOD_MAP is fixed, so it's not really TODO. I usually use TODO for things that needs to be done now and XXX for things that needs to be fixed one day or to explain workarounds. We can put TODO if it seems more suitable to you of course, I'm not sure how widespread it is to use XXX comments

Reply:
Ah yes, that makes sense. TODOs are a bit of a get-out clause aren't they, they can end up just becoming permanent features. So I totally get what you mean. I hadn't heard of XXX, but I may be out of the loop. Though it doesn't have the self-explanatory character of TODO. It's obviously not a big thing, but yes I think it'd be good if it was less obscure.

Reply:
with lsprotocol and pygls 1.0 we no longer need this ugly workaround. I removed it.